### ISSUE TITLE
Introduce forward types to support circular types

### ISSUE THREAD

**Issue by mweststrate** [mobxjs/mobx-state-tree#74]
```javascript
import {getParent, types} from 'mobx-state-tree'

const Store = types.define((self) => types.model({
   todos: types.array(Todo), // without types.define, would be `null` at this point
   subStores: types.array(self) // for recursive types
})

type IStore = typeof Store.Type // already known at compile time

const Todo = types.model({
  title: "",
  getParent(): IStore {
     return getParent(this) as IStore
  }
})

Store.finalizeType() // make sure the lambda is actually called
```

**Comment by mweststrate** [comment]
cc: @mattiamanzati what do you think? It's like #52 but generalized

**Comment by mweststrate** [comment]
Beste name? `forward | define | declare`

**Comment by mattiamanzati** [comment]
@mweststrate Why not let generic types accept a lambda function that resolves to a type? GraphQL does something similar if I remember correclty

```javascript
import {getParent, types} from 'mobx-state-tree'

const Store = types.model({
   todos: types.array(() => Todo), // lambda will be called at create() time
   subStores: types.array(self) // for recursive types
})

type IStore = typeof Store.Type // already known at compile time

const Todo = types.model({
  title: "",
  getParent(): IStore {
     return getParent(this) as IStore
  }
})
```

**Comment by mweststrate** [comment]
Solved through #113
